### ISSUE TITLE
Comparison of ASHP and MSHP performance pre- and post-HPXML

### ISSUE THREAD

**Issue by ejhw** [NREL/OpenStudio-HPXML#1011]
@whiphi92 ran a ResStock precomputed buildstock.csv with 10 samples pre- and post- the move of ResStock to OS-HPXML. We are seeing some unexpected results for MSHPs.
[heat-pump-hpxml-comparison.yml.txt](https://github.com/NREL/OpenStudio-HPXML/files/8325201/heat-pump-hpxml-comparison.yml.txt)

[xlsx with annual COP values and scatterplots is here](https://nrel.sharepoint.com/:x:/s/EUSS-Res/EeQ0-1MpO49JnrspmE2Ba4cB8LVhrJmDfgxG5UdB9eSg9A?e=xUHeOA) 

Couple observations on the annual COP values:

- **QC check: Are COP values reasonable?** Most COP values are much lower than I'd expect, compared to the values in the scatterplots, and when comparing to the rated HSPF values.
   - @whiphi92 questions
     - Is the calculation the same and was used for the hourly scatterplots? 
     - Are these all in MN or multiple climates? 
     - Can you upload the precomputed buildstock.csv somewhere too?
   - We could compare to BEopt results as a gut check.
   - Avg COP of ASHPs increased by 32% with move to HPXML
   - Avg COP of MSHPs decreased by 12% with move to HPXML
 - **QC check: Does the performance of different ASHP and MSHP units roughly follow the HSPF trend?**
   - One would not expect perfect agreement with the HSPF trend, but I'd expect the rank order of COPs to be in the same order as rank order HSPF values.
   - Comparing a trend of "percent change in COP vs HSPF 8.2" to "percent change in HSPF vs the HSPF 8.2 unit", we see that for ASHPs, the trendline decreased a bit with the change to HPXML (slope of 0.8 to 0.7). If MSHPs are included in the trendline, the change to HPXML has a huge impact (slope decreases from 1.18 to 0.18). See plots
   - With HPXML, the 9.5 HSPF MSHP performs _worse_ than the 8.2 HSPF ASHP.

![image](https://user-images.githubusercontent.com/1276021/159506099-68b8b8fc-0699-4c01-8790-c1a2a09d3f5e.png)

![image](https://user-images.githubusercontent.com/1276021/159506687-106b502e-1e2a-497f-af1a-0ad0355a9c4d.png)

Assuming the calcs are correct, this makes me hesitant to use HPXML for upcoming ResStock runs. cc @ekpresent @afontani @shorowit @jonwinkler @joseph-robertson

**Comment by ejhw** [comment]
Hugh's TRM-based bin analysis using NEEP data for Albany got seasonal COP of 2.67. HPXML E+ (both using defaults and modified with NEEP inputs on capacity retention) gets 3.16-3.17, which suggests to me that the calculations of COP above aren't correct.

**Comment by ejhw** [comment]
@whiphi92's COP calculation code:

**pre-HPXML**
`avg_cop = (sum(`HEATING COIL HEATING ENERGY (RES MS HTG COIL) [J]`)/3600000 + sum(electricity_fans_heating_kwh) + sum(`HEATING COIL HEATING ENERGY (RES MS SUPP HEATER) [J]`)/3600000) / ((sum(electricity_heating_kwh) + sum(electricity_heating_supplemental_kwh) + sum(electricity_fans_heating_kwh) + sum(electricity_pumps_heating_kwh))))`

**HPXML**
`avg_cop = (sum(`Load: Heating: Delivered`) + 3.41214 * sum(`End Use: Electricity: Heating Fans/Pumps`)) / (3.41214 * (sum(`End Use: Electricity: Heating`) + sum(`End Use: Electricity: Heating Fans/Pumps`))))`

**Comment by shorowit** [comment]
For the HXPML case, the loads and electricity usage outputs are all in the same units (MBtu/yr), so it is incorrectly applying unit conversions.

**Comment by shorowit** [comment]
Also note that `End Use: Electricity: Heating` in HPXML will contain defrost/crankcase energy use, I can't tell if/how that is included in your pre-HPXML equation.

**Comment by ejhw** [comment]
Thanks Scott. I'd expect `electricity_heating_kwh` to contain defrost/crankcase energy use (though maybe crankcase is in cooling as you mentioned recently).

**Comment by whiphi92** [comment]
@shorowit Hi Scott, I summed values from the timeseries data. Do the units conversions look correct knowing I summed from the timeseries outputs?

**Comment by shorowit** [comment]
I don't feel like I have enough information to know if things are correct here.

I ran the example ductless MSHP HPXML file from this repo:
`openstudio workflow/run_simulation.rb -x workflow/sample_files/base-hvac-mini-split-heat-pump-ductless.xml -o mshp --hourly enduses --hourly loads`
It has a 10.0 HSPF and produces an average heating COP of ~3.3. But if I look at the heating COPs in your spreadsheet, they range from 1.3 to 2.3, and that's with MSHPs that go up to 13.3 HSPF.

**Comment by shorowit** [comment]
Note that the [loads reported by OS-HPXML include HVAC distribution losses](https://openstudio-hpxml.readthedocs.io/en/latest/workflow_outputs.html#annual-building-loads), maybe that's part of the problem?

**Comment by JLReyna** [comment]
I've asked @whiphi92 to also run some samples for Los Angeles, since we need to make a decision on whether to use HPXML for LA100 ASAP.

**Comment by whiphi92** [comment]
As requested by @ejhw, here is a zip file with pre and post HPXML annual upgrade values.
[upgrade_annual_data.zip](https://github.com/NREL/OpenStudio-HPXML/files/8420564/upgrade_annual_data.zip)

**Comment by ejhw** [comment]
New findings/questions:

1. Low COPs in MN appear to be because the HPs are being sized for cooling (~22 kBtu/hr mean), so they rely on supplemental heat for much of the year. @whiphi92 will run with sizing for max load to confirm this.
2. Unmet heating hours average 2500-3000 for ASHPs in HPXML. Unmet hours for pre-HPXML look reasonable. @whiphi92  is looking at timeseries data to investigate why this might be. It does not appear to correlated with foundation type or use of setbacks.
![image](https://user-images.githubusercontent.com/1276021/161849272-8d8ff62a-9b8a-4309-8ac5-69b40525f690.png)

3. MSHP Fan electricity is 5-7x higher in HPXML. Is that expected?
![image](https://user-images.githubusercontent.com/1276021/161849178-78cab4ea-e458-491b-ad37-f07f20df2f53.png)

4. MSHP Heat-CAP-ft curves look different between pre-hPXML and HPXML. Do we know why that is? What is the assumed capacity vs. temperature relationship in HPXML since there are no longer inputs for that?

**Comment by shorowit** [comment]
> What is the assumed capacity vs. temperature relationship in HPXML since there are no longer inputs for that?

Here's what I see [in ResStock v2.5](https://raw.githubusercontent.com/NREL/resstock/v2.5.0/resources/options_lookup.tsv) and [in OS-HPXML](https://github.com/NREL/OpenStudio-HPXML/blob/d3a3bd73ba2c6a398fd722826b608ce5bbe20ba9/HPXMLtoOpenStudio/resources/hvac.rb#L1074-L1119):

  | ResStock v2.5 | OS-HPXML
-- | -- | --
min_cooling_capacity | 0.4 | 0.4
max_cooling_capacity | 1.2 | 1.2
min_cooling_airflow_rate | 200 | 200
max_cooling_airflow_rate | 425 | 425
heating_capacity_offset | 2300 | 0
min_heating_capacity | 0.3 | 0.3
max_heating_capacity | 1.2 | 1.2
min_heating_airflow_rate | 200 | 200
max_heating_airflow_rate | 400 | 400
cap_retention_frac | 0.25 | 0.25
cap_retention_temp | -5 | -5

So, the only difference is the heating_capacity_offset, which was removed in OS-HPXML because it doesn't make sense to assume a fixed offset irrespective of capacity.

**Comment by shorowit** [comment]
>     4. MSHP Heat-CAP-ft curves look different between pre-hPXML and HPXML. Do we know why that is?

I just ran a single ResStock run w/ a MSHP, using develop and the v2.5 branch, and the curves are identical.

develop:

```
Curve:Biquadratic,
  Heat-CAP-fT1,                           !- Name
  1.00292812115385,                       !- Coefficient1 Constant
  -0.010386676170938,                     !- Coefficient2 x
  0,                                      !- Coefficient3 x**2
  0.0259615384615385,                     !- Coefficient4 y
  0,                                      !- Coefficient5 y**2
  0,                                      !- Coefficient6 x*y
  -100,                                   !- Minimum Value of x {BasedOnField A2}
  100,                                    !- Maximum Value of x {BasedOnField A2}
  -100,                                   !- Minimum Value of y {BasedOnField A3}
  100;                                    !- Maximum Value of y {BasedOnField A3}
```

v2.5:

```
Curve:Biquadratic,
  HP_Heat-Cap-fT2,                        !- Name
  1.00292812115385,                       !- Coefficient1 Constant
  -0.010386676170938,                     !- Coefficient2 x
  0,                                      !- Coefficient3 x**2
  0.0259615384615385,                     !- Coefficient4 y
  0,                                      !- Coefficient5 y**2
  0,                                      !- Coefficient6 x*y
  -100,                                   !- Minimum Value of x {BasedOnField A2}
  100,                                    !- Maximum Value of x {BasedOnField A2}
  -100,                                   !- Minimum Value of y {BasedOnField A3}
  100;                                    !- Maximum Value of y {BasedOnField A3}

```

**Comment by ejhw** [comment]
Thanks—really appreciate you looking into this Scott! 

This was the critical piece I was missing:
>  Note that a switchover temperature is not simply a minimum heat pump lockout temperature, we also implement it as a maximum backup heat lockout temperature.

I think the only other outstanding issues are:

1) **Capacity retention**. I wasn't sure what was happening [in OS-HPXML](https://github.com/NREL/OpenStudio-HPXML/blob/d3a3bd73ba2c6a398fd722826b608ce5bbe20ba9/HPXMLtoOpenStudio/resources/hvac.rb#L1074-L1119), so thanks for linking to that code. I have a proposed workaround that I think will allow us to still model any MSHP without exposing new inputs.

2) MSHP Fan electricity is 5-7x higher in HPXML. Maybe this is okay or expected though.

**Comment by ejhw** [comment]
@whiphi92 
Summary of fan electricity change: We changed from a fan power law to a constant W/cfm to avoid issues such as adding an AC to a furnace and seeing furnace fan electricity change. Not ideal...likely over estimates fan energy use for variable speed systems, but we are probably willing to live with it.  

**Next steps:**
- Implement HSPF threshold logic for capacity retention on EUSS branch or resstock develop? (work with Tony/Joe to figure out which place makes sense)
- Confirm unmet hours fix after merging that commit
- Confirm COPs look higher with sizing for max load
- Check fan energy relationship and decide if we're okay with it.

**Comment by whiphi92** [comment]
![ASHP_percent_diff](https://user-images.githubusercontent.com/29104804/163270362-84f1709f-dfb8-441f-b357-ba629d484d20.png)
ASHP % difference from pre-HPXML ResStock.

x-axis is building id

All CA homes are from Los Angeles county, MN homes are generated using the downselect sample for State|MN.

**Comment by ejhw** [comment]
- [ ] @shorowit will look into why ASHPs are sized the same as suppl. in develop
- [ ] @shorowit will look into why MSHPs are sized for cooling load despite input for maxload

**Comment by ejhw** [comment]
Tableau results I was sharing on the call
![image](https://user-images.githubusercontent.com/1276021/163273576-6f335171-5753-4bd2-a469-255f5ec06d6a.png)
[export.xlsx](https://github.com/NREL/OpenStudio-HPXML/files/8484770/export.xlsx)

**Comment by shorowit** [comment]
>     @shorowit will look into why MSHPs are sized for cooling load despite input for maxload in `pre-hpxml`

It's indeed a bug in ResStock v2.5. [This line](https://github.com/NREL/resstock/blob/v2.5.0/resources/measures/HPXMLtoOpenStudio/resources/hvac_sizing.rb#L2970):
`if hvac.HasAirSourceHeatPump`
should be:
`if hvac.HasAirSourceHeatPump || hvac.HasMiniSplitHeatPump`

In the one MN MSHP building I'm looking at, after making that change:
- The heat pump capacity increases by ~170%.
- The heating energy use decreases by ~40%.

**Comment by shorowit** [comment]
> @shorowit will look into why ASHPs are sized the same as suppl. in `develop`

I discussed with @jonwinkler why our HP sizing methodology in OS-HPXML differs from what we did in BEopt/ResStock for the "max load" setting. As best we can remember, it appears to stem from too literal an interpretation of the ANSI 301 Standard, which has confusing language. I'm going to change OS-HPXML to bring back the original methodology. Thank you for bringing this to our attention!

**Comment by shorowit** [comment]
Should we redo runs that incorporate [this bugfix](https://github.com/NREL/OpenStudio-HPXML/issues/1011#issuecomment-1099373729) on the pre-HPXML side and use [this branch](https://github.com/NREL/resstock/pull/896) on the post-HPXML side? That should fix the two checkboxes from [Eric's comment](https://github.com/NREL/OpenStudio-HPXML/issues/1011#issuecomment-1098493912). Unfortunately, OpenStudio 3.4.0-rc1 will be needed for the post-HPXML runs, which means that there will be some changes due to the E+ 22.1 Kiva foundation bugfixes (unless homes w/o foundations are modeled).

**Comment by shorowit** [comment]
> > @shorowit will look into why ASHPs are sized the same as suppl. in `develop`
> 
> I discussed with @jonwinkler why our HP sizing methodology in OS-HPXML differs from what we did in BEopt/ResStock for the "max load" setting. As best we can tell, it appears to stem from too literal an interpretation of the ANSI 301 Standard, which has confusing language. I'm going to change OS-HPXML to bring back the original methodology. Thank you for bringing this to our attention!

Just got some initial feedback from HERS software developers and it appears that we **were** interpreting their language correctly. The methodology for HERS calculations is to just set the heat pump capacity equal to the larger of heating/cooling design loads (which is what we were doing).

That is clearly very different than what ResStock/BEopt were doing. So I think this means that we will need to split the `UseMaxLoadForHeatPumps` into two separate options with different names and different methodologies... 😬

**Comment by ejhw** [comment]
> The methodology for HERS calculations is to just set the heat pump capacity equal to the larger of heating/cooling design loads (which is what we were doing).

That's referring to _nominal_ heat pump capacity, correct?

**Comment by ejhw** [comment]
@shorowit  @joseph-robertson We found a bug/user error causing fan energy to mistakenly be 0.375 W/cfm for MSHPs. 

This is from a MSHP .idf. 
```
Fan:SystemModel,
air source heat pump supply fan, !- Name
...
PowerPerFlow, !- Design Power Sizing Method
794.580001233493, !- Electric Power Per Unit Flow Rate {W/(m3/s)}
1.66667, !- Electric Power Per Unit Flow Rate Per Unit Pressure {W/((m3/s)-Pa)}
0.7, !- Fan Total Efficiency
```
794.58 W/(m3/s) is 0.375 W/cfm.

I realized that in the [options_lookup for HPXML](https://raw.githubusercontent.com/NREL/resstock/develop/resources/options_lookup.tsv), the `heat_pump_type` gets set by the `HVAC Heating Type And Fuel.tsv` instead of the `HVAC Heating Efficiency.tsv` which contains all of the other heat pump arguments. In @whiphi92 's yaml, his upgrade was setting `HVAC Heating Efficiency` to a MSHP option but did not touch `HVAC Heating Type And Fuel`, so the `heat_pump_type` was still being set to `none` and [thus the fan_watts_per_cfm was getting set to ecm_watts_per_cfm (0.375)](https://github.com/NREL/resstock/blob/a43db1a391fb28f4176cbf4d0e81a6fa70171ada/resources/hpxml-measures/HPXMLtoOpenStudio/resources/hpxml_defaults.rb#L1081-L1113) instead of the default value for a ducted or ductless MSHP. 

As a workaround, we can make our yaml upgrade both options at the same time, but I think we should move all three of those arguments from `HVAC Heating Type And Fuel.tsv` to `HVAC Heating Efficiency.tsv` because I could see this mistake happening frequently. `HVAC Heating Type And Fuel.tsv` was always intended to be informational and not actually used to set arguments.

**Comment by shorowit** [comment]
Sounds like a good idea to me. It seems like best practice to assign arguments at the lowest level possible (e.g., avoid assigning arguments for meta-parameters like `HVAC Heating Type And Fuel`).

**Comment by shorowit** [comment]
>     * [ ]  Someone will spot check fan energy against field studies and decide whether to change W/cfm inputs in `develop`

There is some fan power data in https://www.nrel.gov/docs/fy11osti/52175.pdf, Table 2:

![image](https://user-images.githubusercontent.com/5861765/165831537-d1241782-cebd-4ad6-8e52-bb1cb5a9d206.png)

I put this together quickly, but assuming I'm not missing anything, using a constant W/cfm fan power seems reasonable:

![image](https://user-images.githubusercontent.com/5861765/165831684-5aedeb7e-8db7-4063-b019-4530655f3215.png)

![image](https://user-images.githubusercontent.com/5861765/165831694-3452e176-9d32-4ca3-8aa8-1accabb35d56.png)

[mini_split_fan_power.xlsx](https://github.com/NREL/OpenStudio-HPXML/files/8585791/mini_split_fan_power.xlsx)

**Comment by ejhw** [comment]
Thanks @shorowit . That's the paper I looked up too. While those W vs. cfm plots are linear, I don't think that necessarily means they follow a constant W/cfm (not sure if that's what you were implying). The calculated W/cfm values in your spreadsheet vary by up to a factor of 2. I put together a simple calculator to understand how much this matters. 

Based on some rough assumptions about runtime in different speeds, I found that using a constant 0.07 W/cfm instead of varying could result in overpredicting annual fan energy by 30–40%. For MSHPs, that's only ~30 kWh/yr so not a very big deal.

I tried another calculation to estimate the potential impact for variable-speed ASHPs with 0.375 W/cfm (on the ASHP worksheet in the attached xlsx), and found potentially larger impacts of ~500–700 kWh/yr (35–60%). Again, this is based on my assumed number of hours in low/med/high speed, so will change if those hours are different.

[mini_split_fan_power+ew.xlsx](https://github.com/NREL/OpenStudio-HPXML/files/8586114/mini_split_fan_power%2Bew.xlsx)

**Comment by ejhw** [comment]
I was able to find a John Proctor paper with lab testing of various variable-speed blower (as used in multi-stage furnaces, but I think they'd be applicable to VS heat pumps as well).
[Residential Furnace Blower Performance Testing](https://efiling.energy.ca.gov/getdocument.aspx?tn=222296)

I think the most relevant figures are Figures 4.2.1–4.2.10 that show W/CFM vs. CFM. Note that some of the figures are for "multi-speed" and "constant torque" motors, so I mostly looked at the ones labeled "variable speed"; however, all of them are ECMs. 

I think it is correct to focus on the red line which represents the fan curve across the variable motor speeds where ESP is 0.65 at maximum fan speed. The lower 0.5 ESP value gives a shallower slope, but it represents a constant 0.5 IWC ESP at all fan speeds (not realistic because ESP is a function of CFM).
<img width="849" alt="image" src="https://user-images.githubusercontent.com/1276021/165983978-46859858-1e1b-4fc9-8e5e-9e83f1fe03c4.png">

Here's a summary table for the 3 variable speed motors:
![image](https://user-images.githubusercontent.com/1276021/165989495-292ea117-115d-4b6a-b320-9a20b47b4241.png)
Data: [proctor_blower_wcfm_calcs.xlsx](https://github.com/NREL/OpenStudio-HPXML/files/8592429/proctor_blower_wcfm_calcs.xlsx)

To summarize, the power increase factor is 7–10x for 0.65 ESP. So for ASHPs, which seems to agree more with the previous cubic approach. However, I'm fine using the intermediate 4x power reduction approach for now, lacking more definitive data. If anything we'd be underestimating the fan power savings that you'd get with a VSHP. Or we could try running this by Jon or PSD/Hugh. I wonder if Jon is measuring fan power in his ongoing field studies.

cc @whiphi92

**Comment by shorowit** [comment]
Thanks @ejhw. That's a great resource. I expanded your analysis to constant torque and multi-speed systems.

[proctor_blower_wcfm_calcs2.xlsx](https://github.com/NREL/OpenStudio-HPXML/files/8594357/proctor_blower_wcfm_calcs2.xlsx)

So the calculated exponent is:
- Variable-speed: 2.86 - 3.26
- Constant torque: 2.66 - 3.27
- Multi-speed: 2.92 - 2.95

**Comment by shorowit** [comment]
For what it's worth, I looked through the E+ example files and found three types of curves used for multi/variable-speed systems. Here are the three E+ curves compared to linear, squared, and cubed curves.

![image](https://user-images.githubusercontent.com/5861765/166067171-32f54dd3-1671-468b-8b03-3523800e37a7.png)

[ep_fan_power_curves.xlsx](https://github.com/NREL/OpenStudio-HPXML/files/8594635/ep_fan_power_curves.xlsx)

So two of the curves are pretty similar to a squared relationship and one of the curves is pretty similar to a cubed relationship.

**Comment by shorowit** [comment]
Updated the above image/spreadsheet to include the fan curve from ASHRAE 90.1 Table G3.1.3.15, which more closely follows a squared relationship.

**Comment by ejhw** [comment]
> we also implement it as a maximum backup heat lockout temperature.

@shorowit @joseph-robertson Is there currently a way to specify a backup electric heat lock out temperature to avoid use of resistance heat during setback recovery during mild temps ([like this](https://www.energy.wsu.edu/documents/AHT_Electric%20Heat%20Lock%20Out%20on%20Heat%20Pumps%20(2).pdf)) without using the binary switchover temperature control?

**Comment by ejhw** [comment]
Thanks, yes, it is the same request (which I see I also asked about in Oct 2021). Yes, we have single-stage heat pumps with heating setbacks in the baseline stock and it would happen with upgrades as well.

**Comment by shorowit** [comment]
I think this issue has outlived its usefulness as there are no outstanding requests. If any additional items come up, feel free to reopen this or open a new issue.
